### PR TITLE
Adds ability to include related frameworks when using Carthage

### PR DESCRIPTION
**Pull request by rpassis**
Hey all, 

I'm opening this PR for issue #488 early to get the ball rolling and get some feedback on the solution.

The way this is currently designed, and due to our dependency on the file system to evaluate related dependencies, has made it very difficult to test the logic around including related dependencies.

At the moment I am not exactly sure how to get around it and am open to ideas.

Other than that everything else is pretty straightforward and will hopefully make sense but if you have any questions feel free to ask away!

Rog

### CODE REVIEW

**Comment by yonaskolb** [comment]
Awesome work @rpassis! This will be so useful.
I’ll review this properly later, but regarding your testing issue you can write files to disk in a temporary directory during your test. The source generator tests do this. You can also leverage the `carthageBuildPath` option if you want to point to such a temporary directory from a programmatically created spec

**Comment by ellneal** [comment]
Is there no way to use `CarthageKit` to resolve the embedded dependencies rather than reading the file directly?

**Comment by yonaskolb** [comment]
Yeah that's a good point @ellneal. There looks to be `VersionFile` in CarthageKit we can leverage

**Comment by rpassis** [comment]
Just so I understand, are you suggesting that we add Carthage/CarthageKit as a dependency? 

I'm not sure that I could justify that just for reading the version file from disk. There's a lot of baggage that we will need to inherit if we go down that path, plus after a quick glance it looks like the `VersionFile` itself is internal to the framework so we won't be able to access it anyway.

What do you think?

**Comment by yonaskolb** [comment]
Yeah it would mean adding Carthage as a dependency. Advantage is having access to the proper formal and tested version of the file spec. Also means if Carthage updates the file format we can just easily update the dependency.
But as you say @rpassis it looks like it's internal struct, so it's fine to have it in here for now in that case.
Perhaps @ellneal was suggesting though using some of the public apis at a higher level to actually get at each of the binary paths for a dependency. Haven't looked into if that is possible, but from a quick glance it doesn't look like the api is geared towards that.
So all that said I think the approach you've taken here is good.

**Comment by rpassis** [comment]
So I played around a bit with trying to get CarthageKit included as a dependency and it seems like it could be worth a try if there is interest in using it for more than just this particular feature.

I can see potential in, for example, having a project spec that only specifies that it wants to add dependencies from a Cartfile, and `xcodegen` can take care of the rest.

FWIW introducing CarthageKit also means introducing a few other dependencies include ReactiveSwift which in itself is a big deal as it will definitely have a significant impact on how the code base for this project will evolve.

Let me know what you think!

**Comment by ellneal** [comment]
Good points made. I have to admit that my 15 minute foray into doing this did hit a few roadblocks with CarthageKit not building correctly. Maybe something to discuss another time, and not in the interest of implementing this new feature.

**Comment by rpassis** [comment]
@yonaskolb you're right about the Cartfile. 

From what I've seen so far, the goal of the `.version` file is to allow for the build process to use cached dependencies where possible so I can say with a high level of confidence that we will not be able to use it just to find a list of frameworks for a dependency as this just an implementation detail within the build process.

I'll do a bit more digging and if it's a dead end I'll just move along with what's already here, and start working on adding unit tests.

File: Sources/ProjectSpec/CarthageVersionFile.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public enum Key: String, CodingKey, CaseIterable {
+        case iOS
+        case Mac
+        case tvOS
+        case watchOS
+    }
+
+    let _data: [Key: [Reference]]
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Sources/ProjectSpec/CarthageVersionFile.swift]: Fixed in 12a46f20df8da47f928d234d9f673a1f0a19cab3

File: Sources/ProjectSpec/CarthageVersionFile.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
+//
+//  CarthageVersionFile.swift
+//  AEXML
+//
+//  Created by Rogerio de Paula Assis on 1/30/19.
+//
+
+import Foundation
+
+public struct CarthageVersionFile: Decodable {
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Sources/ProjectSpec/CarthageVersionFile.swift]: Fixed in 12a46f20df8da47f928d234d9f673a1f0a19cab3

File: Tests/ProjectSpecTests/CarthageVersionFileTests.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import ProjectSpec
+import Spectre
+import XCTest
+import PathKit
+
+let fixturePath = Path(#file).parent().parent() + "Fixtures"
+
+class CarthageVersionFileTests: XCTestCase {
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Tests/ProjectSpecTests/CarthageVersionFileTests.swift]: Fixed in 12a46f20df8da47f928d234d9f673a1f0a19cab3

File: Sources/XcodeGenKit/PBXProjGenerator.swift [hunk 4]
```diff
@@ -574,25 +574,31 @@ public class PBXProjGenerator {
...
-
-                var platformPath = Path(getCarthageBuildPath(platform: target.platform))
-                var frameworkPath = platformPath + dependency.reference
-                if frameworkPath.extension == nil {
-                    frameworkPath = Path(frameworkPath.string + ".framework")
-                }
-                let fileReference = sourceGenerator.getFileReference(path: frameworkPath, inPath: platformPath)
+                let allDependencies: [Dependency] = {
+                    guard dependency.includeRelated == true else { return [dependency] }
+                    let related = getCarthageRelatedDependencies(for: dependency, in: target.platform)
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: Updated in 26e64a5154a6f32f5ab1cb787b3298883c915439

File: Tests/Fixtures/TestProject/Carthage/Build/.Result.version [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{
+  "Mac" : [
+    {
+      "name" : "RelatedDependencyA",
+      "hash" : "900c4ae616db1680699090ca6da22b0b9468707d4c20089963c0537893ab2774"
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Tests/Fixtures/TestProject/Carthage/Build/.Result.version]: Fixed in ac8224ebfb8e83aa1c49a6564f7aaa7763780f9f

File: Tests/Fixtures/TestProject/Project.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -23,22 +23,34 @@
...
 		03F389DFD30F3CE5A441925C10CAED46 /* MessagesViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C5ABEA2284F13483EFDF7C0E050F0450 /* MessagesViewController.swift */; };
 		03FC73E479EBD3CFF4E55BBC6CC82311 /* MyFramework.h in Headers */ = {isa = PBXBuildFile; fileRef = 6A58A16491CDDF968B0D56DE7EB96D92 /* MyFramework.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		0571B54E79E0EEE6AEFAD4F67AD51C6B /* RelatedDependencyB.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 49AB177F3B11466E69BDE8D35CBF1110 /* RelatedDependencyB.framework */; };
 		097B0B6C198B9A52D4312F11000496FD /* App_watchOS.app in Embed Watch Content */ = {isa = PBXBuildFile; fileRef = A680BE9F68A255B0FB291AE6F0A2B045 /* App_watchOS.app */; settings = {ATTRIBUTES = (RemoveHeadersOnCopy, ); }; };
+		0A5E7A45C7461184B613A920B3C875D2 /* RelatedDependencyE.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = AD4E8DA1D9D161F174789188D3B650DA /* RelatedDependencyE.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Tests/Fixtures/TestProject/Project.xcodeproj/project.pbxproj]: Cool I went with your suggestion, created a dependency fixture and added to the TestProject's Cartfile. At the moment it is pointing to `rpassis/CarthageTestFixture` but feel free to clone that repo to your own and we can update the Cartfile accordingly.

**Comment by rpassis** [comment]
Hey @yonaskolb CI is failing on the project diff here and I just wanted to make sure this is not a pipeline issue. This is what the configuration file looks like at the moment

```
      ...
      - run:
          name: Generate Fixtures
          command: scripts/gen-fixtures.sh
      - run:
          name: Check Fixture Diffs
          command: scripts/diff-fixtures.sh
      - run:
          name: Build Fixtures
          command: scripts/build-fixtures.sh
```

`Generate Fixtures` expects a Carthage/Build folder to be available but `carthage bootstrap` is only run during `Build Fixtures`. Does this need to be changed or am I misunderstanding something?

Thanks!
Rog

File: Sources/XcodeGenKit/CarthageDependencyResolver.swift [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            }
+            let buildPath = project.basePath + "\(self.baseBuildPath)/.\(dependency.reference).version"
+            guard
+                let data = try? buildPath.read(),
+                let carthageVersionFile = try? JSONDecoder().decode(CarthageVersionFile.self, from: data) else {
+                    // No .version file or we've been unable to parse
+                    // so fail gracefully by returning an empty array
+                    return []
+            }
+            return carthageVersionFile.references(for: platform)
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/CarthageDependencyResolver.swift]: What if there is no version file or it fails parsing? Then we still want to return the main dependency

**Comment by yonaskolb** [comment]
Hi @rpassis 

XcodeGen previously didn't require that Carthage dependencies had been built before generation, to be a bit more flexible. What has been added here now does in fact require that carthage has built it's dependencies so that the version file is in place. That should be added to the documentation here.
In terms of resolving this issue I would suggest perhaps checking in the `Carthage/Build` directory and removing the `carthage boostrap` step from the ci script. What do you think?

**Comment by yonaskolb** [comment]
@ellneal any comments on this PR as you were thinking of implementing this yourself?

File: Sources/ProjectSpec/Dependency.swift [hunk 8]
```diff
@@ -10,6 +10,7 @@ public struct Dependency: Equatable {
     public var removeHeaders: Bool = true
     public var link: Bool?
     public var implicit: Bool = false
+    public var includeRelated: Bool?
```
Comment (ellneal) [inline Sources/ProjectSpec/Dependency.swift]: Seen as this is only relevant to carthage dependencies, could it be an associated value to the enum (defaulting to false)?

```swift
public enum DependencyType: Equatable {
    case target
    case framework
    case carthage(includeRelated: Bool)
    case sdk

    public static var carthage: DependencyType {
        return .carthage(includeRelated: false)
    }
}
```

(This works with a few minor changes; tests pass).

File: Sources/XcodeGenKit/CarthageDependencyResolver.swift [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        self.project = project
+    }
+
+    /// Fetches all carthage dependencies for a given target
+    func dependencies(for topLevelTarget: Target) -> [Dependency] {
+        // this is used to resolve cyclical target dependencies
+        var visitedTargets: Set<String> = []
+        var frameworks: [String: Dependency] = [:]
+
+        var queue: [ProjectTarget] = [topLevelTarget]
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Sources/XcodeGenKit/CarthageDependencyResolver.swift]: I didn't write any of this and at first I thought the same as you but if you look closely in the for loop, dependending on the target type we recursively travel project targets and append them to the `queue` array.

File: Sources/XcodeGenKit/CarthageDependencyResolver.swift [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    
+    init(project: Project) {
+        self.project = project
+    }
+
+    /// Fetches all carthage dependencies for a given target
+    func dependencies(for topLevelTarget: Target) -> [Dependency] {
+        // this is used to resolve cyclical target dependencies
+        var visitedTargets: Set<String> = []
+        var frameworks: [String: Dependency] = [:]
```
[the remark this answers is not in the document]
Comment (rpassis) [reply Sources/XcodeGenKit/CarthageDependencyResolver.swift]: In the interest of keeping this PR as short as possible I'd rather treat such refactor as a separate issue.

File: Sources/ProjectSpec/Dependency.swift (hunk 8 above)
Comment (rpassis) [reply Sources/ProjectSpec/Dependency.swift]: great idea, will work on it.

File: Sources/XcodeGenKit/CarthageDependencyResolver.swift (hunk 7 above)
Comment (rpassis) [reply Sources/XcodeGenKit/CarthageDependencyResolver.swift]: Nice catch, I'll work on it.

**Comment by rpassis** [comment]
> Hi @rpassis
> 
> XcodeGen previously didn't require that Carthage dependencies had been built before generation, to be a bit more flexible. What has been added here now does in fact require that carthage has built it's dependencies so that the version file is in place. That should be added to the documentation here.
> In terms of resolving this issue I would suggest perhaps checking in the `Carthage/Build` directory and removing the `carthage boostrap` step from the ci script. What do you think?

This is an interesting one.

We could check in the dependencies but now you have an issue of built frameworks being tied in to a specific Swift toolchain, and the project will break - for example - when we all move to Swift 5.

I thought about just updating the build pipeline so we bootstrap first, then run the generation and diffing. This will work fine in CI but anyone working locally will need to run `carthage bootstrap` before running tests otherwise there will be diffs in the project file as the dependencies required are not available.

I think option #2 is still a better experience, we will just have to make sure it is well documented.

Thoughts?

**Comment by yonaskolb** [comment]
That's why I suggested perhaps having a different fixture that uses related dependencies, not the main one that gets built. Otherwise another solution would be to check in the version files and nothing else, but that might mean Carthage thinks they are already built?

**Comment by rpassis** [comment]
@yonaskolb I'm not sure we should add a different fixtures just for testing that feature, seems like too much overhead for something that should technically work off the main project what do you think?

I favour the option of checking in the .version files as you suggested.

Based on my tests with only the `.version` files in Carthage's build folder, when running `carthage bootstrap --cache-builds` we will still get the `Invalid cache found` message from carthage, and the frameworks will be built as expected.

So this will work with the least amount of friction.

I'm going to push these changes up now to make sure it also works in CI. 

Once that's passed I'll finish writing a couple of outstanding unit tests for the `CarthageDependencyResolver` type and tidy up/rebase this commit.

Let me know what you think!

**Comment by yonaskolb** [comment]
@rpassis sounds good 👍

**Comment by rpassis** [comment]
This is ready for a final pass, thanks @yonaskolb and @ellneal for the feedback and suggestions!

**Comment by yonaskolb** [comment]
Looking good @rpassis. 

I think it will be useful to have a global option as well to opt in across the whole spec, as that is how it will most likely be used.

The property on dependency would be optional, and would override this global option.
The option could be called `includeRelatedCarthageDependencies`. That's a pretty long name though. Thoughts?

**Comment by rpassis** [comment]
Hey @yonaskolb global flag added, let me know if you can see any other issues with the PR.

Also FWIW the diff on `Package.resolved` here is because that file seems to be out of date in master. I took the liberty to commit the up to date version but let me know if you'd prefer to leave it out.

**Comment by yonaskolb** [comment]
Fantastic, thanks @rpassis! Though your committed `Package.resolved` is unnecessary as it includes Carthage and all of its dependencies

**Comment by rpassis** [comment]
Interesting @yonaskolb, I think I don't understand how SwiftPM works in this regard.

I did add Carthage as a dependency to the `Package.swift` file a while ago to test the viability of using Carthage with this PR, and then reverted / discarded all the changes. 

But every time I run `swift test` on a clean copy of my local repository, the `Package.resolved` file gets modified and those dependencies added so I figured it was something that was changed in master and that the resolved file perhaps never got updated.

I just tried `swift test` again with a brand new clone of the repo and only then everything works as expected i.e. the `Package.resolved` file is not modified.

Do you know why that happens?

**Comment by yonaskolb** [comment]
Yeah I think SPM can get into a weird state if you revert changes to the `Package.resolved` file, as it doesn't match what's in `.build`.

Thanks again for implementing this @rpassis! 😄 
I'm gonna add a couple of doc additions in `Usage.md` as well

**Pull request merged**